sycl/doc/SYCL2020-SpecializationConstants.md

```diff
@@ -0,0 +1,939 @@
+   communicating it to DPC++ RT" provides more info on that)
+
+#### 1. Assignment of numeric IDs to specialization constants
+
+This task is achieved by maintaining a map, which holds a list of numeric IDs
+for each encountered symbolic ID of a specialization constant. Those IDs are
+used to identify the specialization constants at SPIR-V level.
+
+As noted above one symbolic ID can have several numeric IDs assigned to it -
+such 1:N mapping comes from the fact that at SPIR-V level, composite
+specialization constants don't have dedicated IDs and they are being identified
+and specialized through their scalar leafs and corresponding numeric IDs.
+
+For example, the following code:
+```
+struct Nested {
+  float a, b;
+};
+struct A {
+  int x;
+  Nested n;
+};
+
+constexpr specialization_id<int> id_int;
+constexpr specialization_id<A> id_A;
+// ...
+  [=](kernel_handler h) {
+    h.get_specialization_constant<id_int>();
+    h.get_specialization_constant<id_A>();
+  }
+```
+
+Will result in the following numeric IDs assignment:
+```
+// since `id_int` is a simple arithmetic specialization constant, we only
+// have a single numeric ID associated with its symbolic ID
+unique_symbolic_id_for_id_int -> { 0 }
+// `id_A` is a composite with three leafs (scalar members, including ones
+// located in nested composite types), which results in three numeric IDs
+// associated with the same symbolic ID
+unique_symbolic_id_for_id_A -> { 1, 2, 3 }
+```
+
+As it is shown in the example above, if a composite specialization constant
+contains another composite within it, that nested composite is also being
+"flattened" and its leafs are considered to be leafs of the parent
+specialization constants. This done by depth-first search through the composite
+elements.
+
+#### 2.a Transformation of LLVM IR to SPIR-V friendly IR form
+
+SPIR-V friendly IR form is a special representation of LLVM IR, where some
+function are named in particular way in order to be recognizable by the SPIR-V
+translator to convert them into corresponding SPIR-V instructions later.
+The format is documented [here][spirv-friendly-ir].
+
+[spirv-friendly-ir]: https://github.com/KhronosGroup/SPIRV-LLVM-Translator/blob/master/docs/SPIRVRepresentationInLLVM.rst
+
+For specialization constant, we need to generate the following constructs:
+```
+template<typename T> // T is arithmetic type
+T __spirv_SpecConstant(int numericID, T default_value);
+
+template<typename T, typename... Elements> // T is composite type,
+// Elements are arithmetic or composite types
+T __spirv_SpecConstantComposite(Elements... elements);
+```
+
+Particularly, `SpecConstantsPass` translates calls to the
+`T __sycl_getScalar2020SpecConstantValue(const char *SymbolicID, const void *DefaultValue, const char *RTBuffer)`
+intrinsic into calls to `T __spirv_SpecConstant(int ID, T default_val)`.
+And for `T __sycl_getComposite2020SpecConstantValue(const chat *SybmolicID, const void *DefaultValue, const char *RTBuffer)`
```

Comment:
```suggestion
And for `T __sycl_getComposite2020SpecConstantValue(const char *SybmolicID, const void *DefaultValue, const char *RTBuffer)`
```

Reply:
Fixed in e88f4027cb8a6385d7db3d2104ca8aea600cdfa9